build-tools/Xamarin.Android.Tools.BootstrapTasks/Xamarin.Android.Tools.BootstrapTasks/CheckApiCompatibility.cs

```diff
@@ -0,0 +1,217 @@
+		[Required]
+		public  string         ApiLevel            	{ get; set; }
+
+		// The last stable api level.
+		[Required]
+		public  string         LastStableApiLevel	{ get; set; }
+
+		// Path to xamarin-android-api-compatibility folder
+		[Required]
+		public  string         ApiCompatContractPath	{ get; set; }
+
+		// Output Path where the assembly was just built
+		[Required]
+		public  string         TargetImplementationPath	{ get; set; }
+
+		// This Build tasks validates that changes are not breaking Api
+		public override bool Execute ()
+		{
+			Log.LogMessage (MessageImportance.High, $"CheckApiCompatibility for ApiLevel: {ApiLevel}");
+
+			// Check to see if Api has a previous Api defined.
+			if (!api_versions.TryGetValue (ApiLevel, out string previousApiLevel)) {
+				Log.LogError ($"Please add ApiLevel:{ApiLevel} to the list of supported apis.");
+				return !Log.HasLoggedErrors;
+			}
+
+			// Get the previous api implementation path by replacing the current api string with the previous one.
+			var previousTargetImplementationPath = TargetImplementationPath.Replace (ApiLevel, previousApiLevel);
+
+			// In case previous api is not defined or directory does not exist we can skip the check.
+			var validateAgainstPreviousApi = !(string.IsNullOrWhiteSpace (previousApiLevel) || !Directory.Exists (previousTargetImplementationPath));
+			if (validateAgainstPreviousApi) {
+
+				// First we check the Api level assembly against the previous api level assembly
+				// i.e.: check api breakages using "the just built V2.dll" against "the just built V1.dll"
+				ValidateApiCompat (previousTargetImplementationPath);
+
+				if (Log.HasLoggedErrors) {
+					return !Log.HasLoggedErrors;
+				}
+			}
+
+			// If Api level is the latest we should also compare it against the reference assembly
+			// located on the external folder. (xamarin-android-api-compatibility)
+			// i.e.: check apicompat using "the just built V2.dll" against V2.dll located on xamarin-android-api-compatibility repo
+			if (ApiLevel == LastStableApiLevel) {
+
+				// Check xamarin-android-api-compatibility reference directory exists
+				var referenceContractPath = Path.Combine (ApiCompatContractPath, "reference");
+				if (!Directory.Exists (referenceContractPath)) {
+					Log.LogMessage (MessageImportance.High, $"CheckApiCompatibility Warning: Skipping reference contract check.\n{referenceContractPath} does not exist.");
+					return !Log.HasLoggedErrors;
+				}
+
+				ValidateApiCompat (referenceContractPath);
+			}
+
+			return !Log.HasLoggedErrors;
+		}
+
+		// Validates Api compatibility between contract (previous version) and implmentation (current version)
+		// We do that by using Microsoft.DotNet.ApiCompat.dll
+		void ValidateApiCompat (string contractPath) {
+
+			var apiCompat = Path.Combine (ApiCompatPath, "Microsoft.DotNet.ApiCompat.exe");
+
+			string content;
+			using (var genApiProcess = new Process()) {
+
+				genApiProcess.StartInfo.FileName = "mono";
```

Comment:
We likely need an OS check here to allow this task to run on Windows, or we'll need to make the target which runs this conditional.

Reply:
@pjcollins we don't install mono on windows?

Reply:
Not as far as I know. We'd want to invoke the .exe file directly on Windows I believe.

Reply:
Added a check if windows vs use mono